datafusion/physical-expr/src/equivalence/projection.rs

```diff
@@ -67,6 +67,9 @@ impl ProjectionMapping {
                             // Conceptually, `source_expr` and `expression` should be the same.
                             let idx = col.index();
                             let matching_input_field = input_schema.field(idx);
+                            if col.name() != matching_input_field.name() {
+                                return Err(DataFusionError::Internal(format!("Input field name {} does not match with the projection expression {}",matching_input_field.name(),col.name())));
+                            }
```

Comment:
This check does not exist before. Now, input field names and projection expressions must match.